src/futures.rs

```diff
@@ -0,0 +1,117 @@
+use std::{future::Future, pin::Pin, task::Context, task::Poll};
+
+use pin_project::pin_project;
+
+use crate::{segment, transaction::Transaction};
+
+/// A trait to make a lifetime scoped reference to a transaction optional
+pub trait OptionalTransaction<'a> {
+    /// Return an optional transaction
+    fn get_transaction(&'a self) -> Option<&'a Transaction>;
+}
+
+impl<'a> OptionalTransaction<'a> for &'a Transaction {
+    fn get_transaction(&'a self) -> Option<&'a Transaction> {
+        Some(self)
+    }
+}
+
+impl<'a> OptionalTransaction<'a> for Option<&'a Transaction> {
+    fn get_transaction(&'a self) -> Option<&'a Transaction> {
+        *self
+    }
+}
+
+/// Extension trait allowing a `std::future::Future` to be instrumented inside a `Segment`
+pub trait Segmented: Sized {
+    /// Instruments this future inside a custom `Segment`
+    fn custom_segment<'a, T>(
+        self,
+        to_trans: &'a T,
+        name: &str,
+        category: &str,
+    ) -> SegmentedFuture<'a, Self>
+    where
+        T: OptionalTransaction<'a>,
+    {
+        match to_trans.get_transaction() {
+            Some(transaction) => SegmentedFuture {
+                inner: self,
+                segment: Some(segment::Segment::custom(transaction, name, category)),
+            },
+            None => SegmentedFuture {
+                inner: self,
+                segment: None,
+            },
+        }
+    }
```

Comment:
Completely down to preference but this could be shortened to...

```suggestion
        SegmentedFuture {
            inner: self,
            segment: to_trans
                .get_transaction
                .map(|t| segment::Segment::custom(t, name, category)),
        }
```

Reply:
Hmm, this suggestion is only replacing the last line even though I highlighted all of the above block! Never mind...